### PR TITLE
Add a hook 'actionValidateOrderAfter', This hook is called after the complete creation of an order

### LINKED ISSUE #23787: Add a hook 'actionValidateOrderAfter', This hook is called after the complete creation of an order
#### Is your feature request related to a problem?

When an order is created, several modules can be called by the validateOrder hook or other hooks.
There are very often problems and an order is not completely created, no status, no confirmation email sent etc... the reasons can be multiple, poorly developed module, unexpected error, call to an api that does not respond.
Linking PrestaShop to an ERP is more and more common and the exchanges must be reliable.
Also sending to an ERP can generate a change of status for the order but since it is not completely created, there can be conflicts.

#### Describe the solution you'd like

This hook allows to create the order completely and then the hook is executed, sending the order to an ERP for example, allows to have the whole order with a status etc. and if the link with the ERP does not work, the order is not "corrupted".
I took the example of link with an ERP because I added this hook on several sites and it is very useful.

#### Alternatives you've considered

the validateOrder hook is very often used but causes a lot of maintenance

#### Additional context

create a module that grafts to this hook, make it so that the hook in the module does not respond.
The command will still be created correctly, unlike ValidateOrder

### PR DESCRIPTION
**Pull request by olecorre**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | add a hoo 'actionValidateOrderAfter', This hook is called after the complete creation of an order
| Type?             | new feature 
| Category?         | BO 
| Fixed ticket?     | #23787
| BC breaks?        | no
| Deprecations?     |  no

When an order is created, several modules can be called by the validateOrder hook or other hooks. 
There are very often problems and an order is not completely created, no status, no confirmation email sent etc... the reasons can be multiple, poorly developed module, unexpected error, call to an api that does not respond.
Linking PrestaShop to an ERP is more and more common and the exchanges must be reliable. 
Also sending to an ERP can generate a change of status for the order but since it is not completely created, there can be conflicts.
This hook allows to create the order completely and then the hook is executed, sending the order to an ERP for example, allows to have the whole order with a status etc. and if the link with the ERP does not work, the order is not "corrupted".
I took the example of laison with an ERP because I added this hook on several sites and it is very useful.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/23786)

### CODE REVIEW

**Comment by olecorre** [comment]
in install-dev/upgrade/sql/1.7.8.0.sql one was missing , in line 12

**Comment by jf-viguier** [comment]
A must have, thanks @olecorre

**Comment by jf-viguier** [comment]
I suggest to add the same arguments as actionValidateOrder :
```
'cart' => $this->context->cart,
'order' => $order,
'customer' => $this->context->customer,
'currency' => $this->context->currency,
'orderStatus' => $order_status
```
The cart, orderStatus can be useful

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
@olecorre Hi Could you rebase your PR ? The 1.7.8.0.sql is in conflict.

File: classes/PaymentModule.php [hunk 1]
```diff
@@ -730,7 +730,9 @@ public function validateOrder(
             if (self::DEBUG_MODE) {
                 PrestaShopLogger::addLog('PaymentModule::validateOrder - End of validateOrder', 1, null, 'Cart', (int) $id_cart, true);
             }
-
+            Hook::exec('actionValidateOrderAfter', [
+                'order' => $order,
+            ]);
             return true;
```
Comment (Progi1984) [inline classes/PaymentModule.php:741]:
```suggestion

            return true;
```

File: install-dev/upgrade/sql/1.7.8.0.sql [hunk 2]
```diff
@@ -22,6 +22,9 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
   (NULL, 'displayAdminGridTableBefore', 'Display before Grid table', 'This hook adds new blocks before Grid component table.', '1'),
   (NULL, 'displayAdminGridTableAfter', 'Display after Grid table', 'This hook adds new blocks after Grid component table.', '1'),
   (NULL, 'displayAdminOrderCreateExtraButtons', 'Add buttons on the create order page dropdown', 'Add buttons on the create order page dropdown', '1'),
+  (NULL, 'actionPresentProductListing', 'Product Listing Presenter', 'This hook is called before a product listing is presented', '1'),
```
Comment (PierreRambaud) [inline install-dev/upgrade/sql/1.7.8.0.sql]: COuld you tell us why you added this? :thinking:

Comment (olecorre) [reply install-dev/upgrade/sql/1.7.8.0.sql]: oups it's an another projet to PR

Comment (olecorre) [reply install-dev/upgrade/sql/1.7.8.0.sql]: How can I cancel it?

File: install-dev/upgrade/sql/1.7.8.0.sql [hunk 3]
```diff
@@ -22,6 +22,9 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
...
+  (NULL, 'actionGetProductPropertiesAfterUnitPrice', 'Product Properties', 'This hook is called after defining the properties of a product', '1'),
```
Comment (PierreRambaud) [inline install-dev/upgrade/sql/1.7.8.0.sql]:
```suggestion
```

**Review by PierreRambaud — APPROVED** [review]

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Oups, just notice you taget 1.7.8.0.sql, but we are in the develop branch, you should use 1.7.9.0.sql instead :)

**Comment by olecorre** [comment]
@PierreRambaud I revert 1.7.8.0.sql and i add sql line in 1.7.9.0.sql ?

**Comment by PierreRambaud** [comment]
> @PierreRambaud I revert 1.7.8.0.sql and i add sql line in 1.7.9.0.sql ?

The file already exists (https://github.com/PrestaShop/PrestaShop/blob/develop/install-dev/upgrade/sql/1.7.9.0.sql) you should rebase your branch :thinking:

File: install-dev/upgrade/sql/1.7.9.0.sql [hunk 4]
```diff
@@ -7,3 +7,5 @@ INSERT INTO `PREFIX_configuration` (`name`, `value`, `date_add`, `date_upd`) VAL
     ('PS_MAIL_DKIM_SELECTOR', '', NOW(), NOW()),
     ('PS_MAIL_DKIM_KEY', '', NOW(), NOW())
 ;
+INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `position`) VALUES
+  (NULL, 'actionValidateOrderAfter', 'New Order', 'This hook is called after the complete creation of an order', '1')
```
Comment (PierreRambaud) [inline install-dev/upgrade/sql/1.7.9.0.sql:11]:
```suggestion

INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `position`) VALUES
  (NULL, 'actionValidateOrderAfter', 'New Order', 'This hook is called after the complete creation of an order', '1');
```

Comment (olecorre) [reply install-dev/upgrade/sql/1.7.9.0.sql:11]: why : ? Rather; no?

Comment (PierreRambaud) [reply install-dev/upgrade/sql/1.7.9.0.sql:11]: haha yes, wrong key :sweat_smile:

**Pull request closed without merging**